### PR TITLE
Support hash partitioning for thanos compactor

### PR DESCRIPTION
**Pull request by yeya24**
Signed-off-by: yeya24 <yb532204897@gmail.com>

This pr adds hash partition support for thanos compactor

| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| API breaks?     | no
| Deprecations?   | no
| Related tickets | none
| License         | Apache 2.0

### What's in this PR?

This pr follows the thanos store gateway hash partition and adds hash partition for thanos compactor.

### Why?

Please see https://thanos.io/tip/thanos/sharding.md/. Thanos compactor also support sharding and it is an important way to scale the compactor.

### Checklist

- [x] Code meets the [Developer Guide](https://github.com/banzaicloud/developer-guide)
- [x] User guide and development docs updated (if needed)
- [x] Related Helm chart(s) updated (if needed)

### To Do

- [x] If the PR is not complete but you want to discuss the approach, list what remains to be done here

### CODE REVIEW

File: thanos/templates/compact-deployment.yaml [hunk 1]
```diff
@@ -1,108 +1,140 @@
...
+        - "--retention.resolution-1h={{ $root.Values.compact.retentionResolution1h }}"
+        - "--block-sync-concurrency={{ $root.Values.compact.blockSyncConcurrency }}"
+        - "--compact.concurrency={{ $root.Values.compact.compactConcurrency }}"
         - "--wait"
-{{ with .Values.compact.extraArgs }}{{ toYaml . | indent 8 }}{{- end }}
+        {{- if $hashPartioning }}
+        - |
+          --selector.relabel-config=
+            - action: hashmod
+              source_labels: ["__block_id"]
```
Comment (yeya24) [inline thanos/templates/compact-deployment.yaml:83]: After some investigation, looks like this configuration doesn't work. I am thinking about making this source label configurable in the helm chart.

This label should identify the blocks from different clusters and group them together.

**Comment by yeya24** [comment]
Hello maintainers, anyone has time to give a review for this pr?

**Comment by yeya24** [comment]
@ahma @tarokkk

File: thanos/templates/compact-deployment.yaml (hunk 1 above)
Comment (pkumarbe) [reply thanos/templates/compact-deployment.yaml]: ok

**Pull request closed without merging**